bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthStoreHandlerImpl.java

```diff
@@ -0,0 +1,531 @@
+                logger.error("Unable to decrypt token, using it directly.", e);
+                accessTokenCache.put(storageKeyAccessToken, accessTokenResponseFromStore);
+            }
+        }
+    }
+
+    /**
+     * Reference is optional -- basically the storage service may be removed during runtime.
+     * When it is removed, the service should still stay up for read-only operations.
+     *
+     * Using default ReferencePolicyOption.RELUCTANT, so it should not be called twice.
+     *
+     * @param storageService
+     */
+    @Reference(cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.DYNAMIC)
+    protected synchronized void setStorageService(StorageService storageService) {
+        if (_storageServiceDoNotUseDirectly == null) {
+            _storageServiceDoNotUseDirectly = storageService;
+            _storeDoNotUseDirectly = storageService.getStorage(STORE_NAME);
+            storageFacade = new StorageFacade(this._storeDoNotUseDirectly);
+        }
+    }
+
+    protected synchronized void unsetStorageService(StorageService storageService) {
+        if (_storageServiceDoNotUseDirectly == storageService) {
+            _storageServiceDoNotUseDirectly = null;
+            storageFacade.close();
+        }
+    }
+
+    /**
+     * Static policy -- dont want to change cipher on the fly!
+     * There may be multiple storage ciphers, choose the one that matches the target (done at activate)
+     *
+     * @param storageCipher
+     */
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE)
+    protected synchronized void setStorageCipher(StorageCipher storageCipher) {
+        // keep all ciphers
+        allAvailableStorageCiphers.add(storageCipher);
+    }
+
+    protected synchronized void unsetStorageCipher(StorageCipher storageCipher) {
+        allAvailableStorageCiphers.remove(storageCipher);
+        if (this.storageCipher.isPresent() && this.storageCipher.get() == storageCipher) {
+            this.storageCipher = Optional.empty();
+        }
+    }
+
+    private boolean isExpired(Date lastUsed) {
+        if (lastUsed == null) {
+            return false;
+        }
+        return System.currentTimeMillis() - lastUsed.getTime() > TimeUnit.MILLISECONDS.convert(EXPIRE_DAYS,
+                TimeUnit.DAYS);
+    }
+
+    private class StorageEntryRemovalListener implements RemovalListener<String, AccessTokenResponse> {
+        @Override
+        public void onRemoval(RemovalNotification<String, AccessTokenResponse> notification) {
+            String handle = notification.getKey();
+            // remove token, last use, and configuration too
+            storageFacade.removeByHandle(handle);
+        }
+    }
+
+    /**
+     * This is designed to simplify all the locking required for the store,
+     * which may get removed/ destroyed/ unavailable due to OSGI policy.
+     */
+    private class StorageFacade implements AutoCloseable {
+        private volatile Storage<String> _storage;
```

Comment:
variable name with _. Not sure if it is actually followed in ESH.

Reply:
Personally, I hate fields with _ at the beginning.   

I have named it this way as an special pattern so it is not confused with other store/ storage variables, nor will they be confused with variables outside of the inner class scope.  

Added comments.